discovery-provider/src/queries/search_es.py

```diff
@@ -0,0 +1,412 @@
+        if current_user_id:
+            response["followed_users"] = pluck_hits(mfound["responses"].pop(0))
+
+    if do_playlists:
+        response["playlists"] = pluck_hits(mfound["responses"].pop(0))
+        if current_user_id:
+            response["saved_playlists"] = pluck_hits(mfound["responses"].pop(0))
+
+    if do_albums:
+        response["albums"] = pluck_hits(mfound["responses"].pop(0))
+        if current_user_id:
+            response["saved_albums"] = pluck_hits(mfound["responses"].pop(0))
+
+    # hydrate users, saves, reposts
+    item_keys = []
+    user_ids = set()
+    if current_user_id:
+        user_ids.add(current_user_id)
+
+    # collect keys for fetching
+    for k in [
+        "tracks",
+        "saved_tracks",
+        "playlists",
+        "saved_playlists",
+        "albums",
+        "saved_albums",
+    ]:
+        for item in response[k]:
+            item_keys.append(item_key(item))
+            user_ids.add(item.get("owner_id", item.get("playlist_owner_id")))
+
+    # fetch users
+    users_by_id = {}
+    current_user = None
+
+    if user_ids:
+        users_mget = esclient.mget(index=ES_USERS, ids=list(user_ids))
+        users_by_id = {d["_id"]: d["_source"] for d in users_mget["docs"] if d["found"]}
+        if current_user_id:
+            current_user = users_by_id.get(str(current_user_id))
+        for id, user in users_by_id.items():
+            users_by_id[id] = populate_user_metadata_es(user, current_user)
+
+    # fetch followed saves + reposts
+    # TODO: instead of limit param (20) should do an agg to get 3 saves / reposts per item_key
+    (follow_saves, follow_reposts) = fetch_followed_saves_and_reposts(
+        current_user_id, item_keys, 20
+    )
+
+    # tracks: finalize
+    for k in ["tracks", "saved_tracks"]:
+        tracks = response[k]
+        hydrate_user(tracks, users_by_id)
+        hydrate_saves_reposts(tracks, follow_saves, follow_reposts)
+        response[k] = transform_tracks(tracks, users_by_id, current_user)
+
+    # users: finalize
+    for k in ["users", "followed_users"]:
+        users = drop_copycats(response[k])
+        response[k] = [
+            extend_user(populate_user_metadata_es(user, current_user)) for user in users
+        ]
+
+    # playlists: finalize
+    for k in ["playlists", "saved_playlists"]:
+        playlists = response[k]
+        hydrate_saves_reposts(playlists, follow_saves, follow_reposts)
+        hydrate_user(playlists, users_by_id)
+        response[k] = [
+            extend_playlist(populate_track_or_playlist_metadata_es(item, current_user))
+            for item in playlists
+        ]
+
+    return response
+
+
+def base_match(search_str: str, operator="or"):
+    return [
+        {
+            "multi_match": {
+                "query": search_str,
+                "fields": [
+                    "suggest",
+                    "suggest._2gram",
+                    "suggest._3gram",
```

Comment:
just curious does 2gram give us a lot in addition to having trigrams?

Reply:
I was a bit confused about how `search_as_you_type` worked at first.

By default it uses the default tokenizer (words) and then builds ngrams with those word tokens.  In the past I've always used `ngrams` to describe letter couplets and `shingles` to describe word couplets... but in current config I think these ngrams fields are working with word tokens

So for example with `you took that photo`:
* 2grams: `you took` `took that` `that photo`
* 3grams: `you took that` `took that photo`

This still helps with phrase matching because the default `suggest` will match if any of the words appear in any order... but additional matches on `2grams` and `3grams` will happen if words appear in same order in the query and document, so more relevant results will rank higher.  At least that's my current understanding... the `search_as_you_type` docs are a bit sparse haha.

We were considering if `search_as_you_type` was giving us much over just using a standard analyzer, but it seemed to be working well enough, so we kept using it.  Since these fields are short (a handful of terms) the overhead of building these 2gram and 3gram subfields is negligible.

But might revisit in future mapping changes.

Reply:
yea that's how i understand it too and why i think `track title` `artist` vs `artist` `track title` ordering slightly matters.